src/v/io/tests/pager_test.cc

```diff
@@ -0,0 +1,335 @@
+                 ++it, ++pit) {
+                auto& prev = *pit;
+                auto& curr = *it;
+                EXPECT_EQ(prev->offset() + prev->data().size(), curr->offset());
+            }
+
+            /*
+             * page range must start at or before the target offset. the result
+             * should be minimal, meaning the first page should also contain the
+             * target offset, not merely be ordered before the target. same for
+             * the end of the range.
+             */
+            EXPECT_LE(pages.front()->offset(), offset);
+            EXPECT_LT(
+              offset, pages.front()->offset() + pages.front()->data().size());
+            EXPECT_LT(pages.back()->offset(), offset + len);
+
+            // no page should ever start after the end of the file
+            EXPECT_LT(pages.back()->offset(), file_size());
+
+            /*
+             * the total bytes returned may exceed the read length in order to
+             * cover the range (e.g. len=1 will still need to return a full
+             * page). in this case it should only be the final page that
+             * accounts for the excess.
+             */
+            const auto total = std::accumulate(
+              pages.begin(), pages.end(), size_t{0}, [](size_t acc, auto page) {
+                  return acc + page->data().size();
+              });
+            if (total > len) {
+                EXPECT_LT(total - pages.back()->data().size(), len);
+            }
+
+            /*
+             * the ending offset of the range formed by the intersection of the
+             * eof, the requested range, and the pages returned from read().
+             */
+            const auto end_offset = std::min(
+              std::min(file_size(), offset + len),
+              pages.back()->offset() + pages.back()->data().size());
+
+            // total requested data read (size of the intersection)
+            const auto read_size = end_offset - offset;
+
+            /*
+             * an input stream over the intersection of the returned pages from
+             * filemap::read and the requested [offset, len) range.
+             */
+            seastar::input_stream<char> input1(
+              seastar::data_source(std::make_unique<io::paging_data_source>(
+                pages, io::paging_data_source::config{offset, read_size})));
+
+            /*
+             * an input stream over a view of the input seed data constrained to
+             * what was returned by pager::read.
+             */
+            seastar::input_stream<char> input2(
+              seastar::data_source(std::make_unique<memory_data_source>(
+                data().share(offset, read_size))));
+
+            EXPECT_TRUE(EqualInputStreams(input1, input2));
+        }
+    }
+}
+
+INSTANTIATE_TEST_SUITE_P(
+  Pager,
+  PagerTest,
+  ::testing::Combine(
+    ::testing::Bool(), ::testing::ValuesIn(test_case_sizes())));
+
+TEST_P(AppendPagerTest, Append) {
```

Comment:
It would be great to have a test that tested appends and concurrent reads, especially with a cache that evicts aggressively

Reply:
oh yeh good call. we'll need to add locking for the append/read case you mention, and also for eviction. none of that is here though in this version that assumes no concurrency.